---
title: "Application_mouse"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    df_print: paged
date: '`r format(Sys.time(), "%d %B, %Y at %H:%M:%S")`'
---

```{r}
rm(list = ls())
gc()
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, warning = TRUE, message = FALSE, fig.align = "center", fig.ext='png', fig.path = "Application_mouse/", cache = FALSE)
```

```{r}
dist_trees = "euclidean"
agreg_trees = "ward.D2"
agreg_mfa = "ward.D2"
```

# Packages

```{r}
library(MouseGastrulationData)
library(SingleCellExperiment)
library(MultiAssayExperiment)
library(SummarizedExperiment)
library(scater)
library(scran)

library(factoextra)
library(FactoMineR)

library(kableExtra)
library(tidyverse)

library(reshape2)
library(ggplot2)
library(gridExtra)
library(dendextend)
library(ggpubr)

library(corrplot)

library(parallel)
library(parallelDist)

library(dynamicTreeCut)
```

```{r}
set.seed(1992)
```

# Functions

```{r}
source("Functions.R")
```

# Loading data

https://bioconductor.org/packages/release/data/experiment/vignettes/MouseGastrulationData/inst/doc/MouseGastrulationData.html

```{r}
AtlasSampleMetadata[AtlasSampleMetadata$stage == "E6.5",]
```

```{r}
sce <- EmbryoAtlasData(samples = AtlasSampleMetadata$sample[AtlasSampleMetadata$stage == "E6.5"])
```

```{r}
sce
```

```{r}
counts(sce)[6:9, 1:3]
```

```{r}
head(sizeFactors(sce))
```

```{r}
head(rowData(sce))
```

```{r}
head(colData(sce))
```

The vignette recommends to use the "normalize" function of R package scater. The package indicates the function is deprecated and the user needs to use "logNormCounts".

```{r}
sce_norm = scater::logNormCounts(sce[,!colData(sce)$stripped])
```

```{r}
logCounts = assay(sce_norm, "logcounts")
dim(logCounts)
```

```{r}
metadata = colData(sce)
metadata = metadata[!metadata$stripped,]
```

```{r}
metadata = metadata[order(metadata$cell),]
logCounts = logCounts[,order(colnames(logCounts))]
```

```{r}
identical(metadata$cell, colnames(logCounts), attrib.as.set = FALSE)
```

## Data transformation

```{r}
Y_genes = read.table("Y_genes_mouse.txt", header = TRUE, sep = ",")
```

```{r}
logCounts = logCounts[which(Matrix::rowSums(logCounts)!=0),which(Matrix::colSums(logCounts)!=0)]
logCounts = logCounts[!rownames(logCounts) == "ENSMUSG00000086503",] # sex gene, found in  https://github.com/MarioniLab/EmbryoTimecourse2018/blob/master/analysis_scripts/atlas/core_functions.R function getHVGs
# also exclude genes ychr
logCounts = logCounts[!rownames(logCounts) %in% Y_genes$Gene.stable.ID, ]
```

```{r}
metadata = metadata[metadata$cell%in%colnames(logCounts),]
```

```{r}
logCounts = logCounts[which(Matrix::rowSums(logCounts)>(10^-3)*ncol(logCounts)),]
```

```{r}
dim(logCounts)
```

## Most variable genes

https://ivanek.github.io/analysisOfGenomicsDataWithR/09_scRNAseq_intro_html.html#9_mean-variance_relationship

```{r}
fTV = modelGeneVar(as.matrix(logCounts))
fit.trend <- metadata(fTV)
```

```{r}
plot(fit.trend$mean, fit.trend$var, xlab = "Mean of log-expression",  ylab = "Variance of log-expression")
curve(fit.trend$trend(x), col = "dodgerblue", add = TRUE, lwd = 2)
```

```{r}
head(fTV[order(fTV$bio, decreasing = TRUE), ])
```

```{r}
fTV[rownames(fTV) == "ENSMUSG00000086503",]
```


```{r}
prop = c(10^-6, 10^-3, 0.01, 0.05, 0.1)
kable(rbind(prop, sapply(prop, FUN = function(x) sum(fTV$FDR<x)))) %>% 
  kable_styling(full_width = FALSE) 
```

```{r}
logCounts_FDR = logCounts[rownames(logCounts)%in%rownames(fTV[fTV$FDR<0.1,]),]
```

```{r}
dim(logCounts_FDR)
```

```{r}
save(logCounts_FDR, fTV, file = "mouseDenis.RData")
```


## CrÃ©ation liste par type de cellules

```{r}
unique(metadata$celltype)
```

```{r}
listLogCounts = by(t(as.matrix(logCounts_FDR)), INDICES = metadata$celltype, FUN = function(tab) tab)
```

```{r}
listLogCounts = listLogCounts[sapply(listLogCounts, nrow)>1]
```

```{r}
kable(data.frame("Group" = names(listLogCounts), "Nb_Samples" = sapply(listLogCounts, nrow)),row.names = FALSE) %>% kable_styling()
```

```{r}
sapply(listLogCounts, FUN = function(tab) sum(colSums(tab)==0))
```

```{r}
listLogCounts = lapply(listLogCounts, FUN = function(tab) tab[order(rownames(tab)), order(colnames(tab))])
```

```{r}
sapply(listLogCounts, FUN = function(tab) sapply(listLogCounts, FUN = function(tab2) identical(colnames(tab), colnames(tab2), attrib.as.set = FALSE)))
```

```{r}
Reduce(intersect, lapply(listLogCounts, FUN = function(tab) which(colSums(tab) == 0)))
```


# Tree integration

## Building hierarchical clustering and retrieving coordinates

```{r}
hc_list = lapply(listLogCounts, FUN = function(x) hclust(dist(t(x), method = dist_trees), method = agreg_trees))
```

```{r}
cop_list = lapply(hc_list, cophenetic)
```

```{r}
mds_list = lapply(cop_list, cmdscale, k = nrow(as.matrix(cop_list[[1]]))-1)
```

<!-- Note: negative values are encountered, very close to 0 (e-11 at most) -->

## MFA

```{r}
MFA_res = MFA(do.call("cbind", mds_list),
              group = unlist(lapply(mds_list, ncol)),
              name.group = names(mds_list),
              type = rep("c", length(mds_list)), ncp = Inf, graph = FALSE)
```

```{r}
fviz_eig(MFA_res, addlabels = TRUE)
kable(data.frame(round(MFA_res$group$RV,2)))%>% kable_styling(full_width = FALSE)
```

### Heatmap Contribution

```{r, fig.height = 5, fig.width=12}
corrplot(MFA_res$group$contrib[,1:50], method = "circle", mar = c(0, 0, 1.5, 0), bg = "antiquewhite", diag = TRUE, title = "Contribution, 50 first axes", is.corr = FALSE)
```

## Cell type clustering

```{r}
hc_samples = hclust(dist(MFA_res$group$coord), method = agreg_mfa)
```

```{r}
DTC_AFM = dynamicTreeCut::cutreeDynamic(hc_samples, minClusterSize = 1, distM = as.matrix(dist(MFA_res$group$coord)))
```

```{r}
names(DTC_AFM) = names(mds_list)
```


```{r}
hc_samples_dend = branches_attr_by_clusters(as.dendrogram(hc_samples), clusters = (as.numeric(DTC_AFM)+1)[hc_samples$order])
labels_colors(hc_samples_dend) = get_leaves_branches_col(hc_samples_dend)
```

```{r, fig.height = 6, fig.width = 6.1}
# png(file = "hc_samples.png", width = 1.5*9.5, height = 1.5*9.5, units = "cm", res = 300)
ggdend = fviz_dend(hc_samples_dend, horiz = TRUE, ggtheme = theme_bw(),main = "Hierarchical Clustering of cell types",  cex = 1,
          labels_track_height = 0.2, lwd = 1)  + theme(text = element_text(size = 15), axis.title = element_text(size = 17))
ggdend
# dev.off()
```

## Factorial Maps

```{r, fig.height = 3.5 , fig.width = 12}
gg_tab = data.frame(MFA_res$group$coord, "Data" = rownames(MFA_res$group$coord), "Classification" = as.factor(DTC_AFM))

xlim_max = max(MFA_res$group$coord[,1]) + 0.05
ylim_max = max(MFA_res$group$coord[,2]) + 0.05

cols = c("1" = "#00BA38", "2" = "#F8766D", "3" = "#619CFF")

gg1 = ggplot(gg_tab, aes(x = Dim.1, y = Dim.2, color = Classification)) + 
  geom_point() + geom_text(aes(label=rownames(gg_tab)),hjust=1, vjust=1) +
  ggtitle("Axes 1/2") + theme_bw() + xlim(c(-xlim_max,xlim_max)) + ylim(c(-0.05, ylim_max)) +
  xlab(paste0("Dim 1 - ", round(MFA_res$eig[1,2], 2), "%")) + ylab(paste0("Dim 2 - ", round(MFA_res$eig[2,2], 2), "%")) +
  theme(text = element_text(size = 15), axis.title = element_text(size = 15)) +
  scale_color_manual(values = cols) + theme(legend.position = "none")
gg2 = ggplot(gg_tab, aes(x = Dim.2, y = Dim.3, color = Classification)) + 
  geom_point() + geom_text(aes(label=rownames(gg_tab)),hjust=1, vjust=1) +
  ggtitle("Axes 2/3") + theme_bw() + xlim(c(-xlim_max,xlim_max)) + ylim(c(-0.05, ylim_max)) +
  xlab(paste0("Dim 2 - ", round(MFA_res$eig[2,2], 2), "%")) + ylab(paste0("Dim 3 - ", round(MFA_res$eig[3,2], 2), "%")) +
  theme(text = element_text(size = 15), axis.title = element_text(size = 15))+
  scale_color_manual(values = cols) + theme(legend.position = "none")
gg3 = ggplot(gg_tab, aes(x = Dim.3, y = Dim.4, color = Classification)) + 
  geom_point() + geom_text(aes(label=rownames(gg_tab)),hjust=1, vjust=1) +
  ggtitle("Axes 3/4") + theme_bw() + xlim(c(-xlim_max,xlim_max)) + ylim(c(-0.05, ylim_max)) +
  xlab(paste0("Dim 3 - ", round(MFA_res$eig[3,2], 2), "%")) + ylab(paste0("Dim 4 - ", round(MFA_res$eig[4,2], 2), "%")) +
  theme(text = element_text(size = 15), axis.title = element_text(size = 15))+
  scale_color_manual(values = cols)+ theme(legend.position = "none")
gg4 = ggplot(gg_tab, aes(x = Dim.4, y = Dim.5, color = Classification)) + 
  geom_point() + geom_text(aes(label=rownames(gg_tab)),hjust=1, vjust=1) + 
  ggtitle("Axes 4/5") + theme_bw() + xlim(c(-xlim_max,xlim_max)) + ylim(c(-0.05, ylim_max)) +
  xlab(paste0("Dim 4 - ", round(MFA_res$eig[4,2], 2), "%")) + ylab(paste0("Dim 5 - ", round(MFA_res$eig[5,2], 2), "%")) +
  theme(text = element_text(size = 15), axis.title = element_text(size = 15))+
  scale_color_manual(values = cols) + theme(legend.position = "none")
# grid.arrange(gg1, gg2, gg3, gg4, ncol = 4)
```

```{r}
ggall = ggarrange(gg1, gg2, gg3, gg4, ncol = 2, nrow = 2, common.legend = TRUE, legend="bottom")
ggall
```



## RV heatmap

```{r, fig.height = 6, fig.width = 6.1}
# postscript(file = "RV_AFM.eps", width = 8, height = 8, horizontal = FALSE, onefile = FALSE, paper = "special")
corrplot2(MFA_res$group$RV[-ncol(MFA_res$group$RV),-ncol(MFA_res$group$RV)], method = "circle", mar = c(0, 0, 0, 0), bg = "gray88", diag = FALSE, tl.cex = 1.2, tl.col = "black", is.corr = FALSE,
         order = hc_samples_dend,
        coef_thresh = 0.44, addCoef.col = "gray98") 
corrPlot = recordPlot()
# dev.off()
```

```{r, fig.height = 6, fig.width = 6.1}
# png(file = "RV_AFM.png", width = 1.5*9.5, height = 1.5*9.5, units = "cm", res = 300)
corrplot2(MFA_res$group$RV, method = "circle", mar = c(0, 0, 0, 0), bg = "gray88", diag = FALSE, tl.cex = 1.2, tl.col = "black", is.corr = FALSE,
         order = hc_samples_dend,
        coef_thresh = 0.44, addCoef.col = "gray98") 
# dev.off()
```

## Consensus tree on global MFA

```{r, fig.height = 5, fig.width = 10}
g1 = fviz_mfa_ind(MFA_res, geom = "point")
g2 = fviz_mfa_ind(MFA_res, geom = "point", axes = c(2,3))
ggarrange(g1, g2, ncol = 2)
```


```{r}
hc_gMFA = hclust(dist(MFA_res$ind$coord, method = "euclidean"), method = "ward.D2")
```

```{r}
hc_gMFA2 = hc_gMFA
hc_gMFA2$labels = NA
par(mar = c(0.5, 2.5, 2.5, 0.1))
plot(as.dendrogram(hc_gMFA2), main = "Hierarchical clustering of genes \n global MFA")
```

```{r}
# png(filename = "mouse_gMFA_hc.png", width = 1.5*15, height = 1.5*7.5, res = 300, units = c("cm"))
hc_gMFA2 = hc_gMFA
hc_gMFA2$labels = NA
par(mar = c(0.5, 2.5, 2.5, 0.1))
plot(as.dendrogram(hc_gMFA2), main = "Hierarchical clustering of genes \n global MFA")
# dev.off()
```

## Consensus tree on MFA subgroups

```{r}
names(DTC_AFM) = names(mds_list)
```

```{r}
mds_sublists = lapply(1:max(DTC_AFM), FUN = function(i) mds_list[names(mds_list)%in%names(DTC_AFM[DTC_AFM == i])])
```

```{r}
mfa_sublists = lapply(mds_sublists, FUN = function(liste){
  MFA(do.call("cbind", liste),
              group = unlist(lapply(liste, ncol)),
              name.group = names(liste),
              type = rep("c", length(liste)), ncp = Inf, graph = FALSE)
})
```

```{r}
hc_cs_sub = lapply(mfa_sublists, FUN = function(MFARes) hclust(dist(MFARes$ind$coord), method = "ward.D2"))
```

```{r}
par(mar = c(0.5, 2.5, 2.5, 0.1))
res = lapply(1:length(hc_cs_sub), FUN = function(i){
  hc = hc_cs_sub[[i]]
  hc$labels = NA
  plot(as.dendrogram(hc), main = paste0("Consensus Tree - Group ", i)) 
})
```


# Figures (MDS+MFA)

```{r}
library(ggplot2)
library(cowplot)
```

## Clustering/RV/Maps

```{r}
ggsave(filename = "Figure8.pdf",
       plot =  ggarrange(ggdraw(ggdend, clip = "on", xlim = c(-0.05, 1)),  ggdraw(ggall, clip = "on"), ncol = 2, labels = c("A)", "B)")),
       width = 2*15, height = 2*6.5, units = c("cm"))
```

## Consensus trees

```{r}
par(mar = c(0.5, 2.5, 2.5, 0.1))

hc = hc_gMFA
hc$labels = NA

plot(as.dendrogram(hc), main = "Consensus Tree \n global MFA", cex = 1, cex.axis = 1, cex.main = 1) 
p1 = recordPlot()

hc = hc_cs_sub[[1]]
hc$labels = NA
plot(as.dendrogram(hc), main = paste0("Consensus Tree \n Group ", 1), cex = 1, cex.axis = 1, cex.main = 1)  
p2 = recordPlot()

hc = hc_cs_sub[[2]]
hc$labels = NA
plot(as.dendrogram(hc), main = paste0("Consensus Tree  \n Group ", 2), cex = 1, cex.axis = 1, cex.main = 1) 
p3 = recordPlot()

hc = hc_cs_sub[[3]]
hc$labels = NA
plot(as.dendrogram(hc), main = paste0("Consensus Tree  \n Group ", 3), cex = 1, cex.axis = 1, cex.main = 1) 
p4 = recordPlot()
```

```{r, fig.width = 15, fig.height=5}
ggsave(filename = "Figure9.png",  
        plot =    ggarrange(ggdraw(p1, clip = "on"), 
          ggdraw(p2, clip = "on"), 
          ggdraw(p3, clip = "on"), 
          ggdraw(p4, clip = "on"), ncol = 4, labels = ''), width = 2*15, height = 2*5, units = c("cm"))
```

# Multiple Kernel integration

```{r}
library(mixKernel)
```

## Transformation distance matrices to kernels

```{r}
list_kernel_coph_mat = lapply(cop_list, FUN = function(mat) -double_centering(mat^2)/2)
```

```{r}
list_kernel_coph = lapply(list_kernel_coph_mat, FUN = function(mat){
  kern = list(kernel = mat, kernel.fun = "cophenetic")
  class(kern) = "kernel"
  kern
})
```

## Kernels combination

```{r}
list_args = list_kernel_coph
list_args$method = "full-UMKL"
combKern = do.call(combine.kernels, list_args)
```

```{r}
round(combKern$weights, 3)
```

```{r}
mat_sim = similarities(list_kernel_coph)
rownames(mat_sim) = colnames(mat_sim) = names(list_kernel_coph)
```

```{r}
mat_dist = DistFromSim(mat_sim)
```

```{r}
hc_samples = hclust(as.dist(mat_dist), method = "complete")
```


```{r}
DTC_Kern = dynamicTreeCut::cutreeDynamic(hc_samples, minClusterSize = 1, distM = as.matrix(as.dist(mat_dist)))
k_cut = max(DTC_Kern)
groups = cutree(hc_samples, k = k_cut)
```


```{r, fig.height = 5, fig.width = 5}
corrplot2(mat_sim, method = "circle", mar = c(0, 0, 0, 0), bg = "gray85", diag = FALSE, tl.cex = 1, tl.col = "black", 
         is.corr = FALSE, addrect = k_cut, order = hc_samples, addCoef.col = "antiquewhite", coef_thresh = 0.5)
corrPlot = recordPlot()
```

```{r}
hc_samples_dend = branches_attr_by_clusters(as.dendrogram(hc_samples), clusters = (as.numeric(DTC_Kern)+1)[hc_samples$order])
labels_colors(hc_samples_dend) = get_leaves_branches_col(hc_samples_dend)
```

```{r, fig.height = 6, fig.width = 6.5}
# png(file = "hc_samples_kernels.png", width = 1.5*9.5, height = 1.5*9.5, units = "cm", res = 300)
ggdend = fviz_dend(hc_samples_dend, horiz = TRUE, ggtheme = theme_bw(),main = "Hierarchical Clustering of cell types",  cex = 1,
          labels_track_height = 0.7, lwd = 1)  + theme(text = element_text(size = 15), axis.title = element_text(size = 17))
# ggdend
# dev.off()
ggdend
```

# Figures (kernels)

```{r}
ggsave(filename = "Supplementary2.pdf",
       plot =  ggarrange(ggdraw(fviz_dend(hc_samples_dend, horiz = TRUE, ggtheme = theme_bw(),main = "Hierarchical Clustering of cell types",  cex = 1,
                                          labels_track_height = 0.7, lwd = 1)  + theme(text = element_text(size = 15), axis.title = element_text(size = 17)), clip = "on", xlim = c(-0.05, 1)),  ggdraw(corrPlot, clip = "on"), ncol = 2, labels = c("A)", "B)")),
       width = 2*15, height = 2*6.5, units = c("cm"))
```

# InfoSession

```{r}
sessionInfo()
```
